### PR TITLE
Waf active rules

### PR DESCRIPTION
**Pull request by oscarDovao**
This PR contains the work which enables the creation/update/deletion of a WAF's active rules.
A function for upsert and delete has been added, this function can be use for batch process, however the batching has to be managed by the caller.

### CODE REVIEW

File: fastly/errors.go [hunk 1]
```diff
@@ -93,7 +93,7 @@ var ErrMissingIP = errors.New("Missing requried field 'IP'")
 // requires a "EventID" key, but one was not set
 var ErrMissingEventID = errors.New("Missing required field 'EventID'")
 
-// ErrMissingWafID is an error that is returned when an input struct
+// ErrMissingWAFID is an error that is returned when an input struct
```
Comment (phamann) [inline fastly/errors.go]:
```suggestion
// ErrMissingWAFID is an error that is returned when an input struct.
```

File: fastly/errors.go [hunk 2]
```diff
@@ -105,6 +105,10 @@ var ErrMissingWAFVersionNumber = errors.New("Missing required field 'WAFVersionN
 // requires a "WAFVersionID" key, but one was not set.
 var ErrMissingWAFVersionID = errors.New("Missing required field 'WAFVersionID'")
 
+// ErrMissingWAFActiveRuleList is an error that is returned when an input struct
+// requires a list of WAF active rules, but it is empty
```
Comment (phamann) [inline fastly/errors.go]:
```suggestion
// requires a list of WAF active rules, but it is empty.
```

File: fastly/waf_active_rule.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
+package fastly
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/google/jsonapi"
+	"io"
+	"reflect"
+	"strconv"
+	"time"
```
Comment (phamann) [inline fastly/waf_active_rule.go]: Please separate standard library imports from others.

File: fastly/waf_active_rule.go [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+// WAFActiveRuleResponse represents a list of active rules - full response.
+type WAFActiveRuleResponse struct {
+	Items []*WAFActiveRule
+	Info  infoResponse
+}
+
+// ListWAFActiveRulesInput used as input for listing a WAF's active rules.
+type ListWAFActiveRulesInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_active_rule.go [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+	}
+	return &WAFActiveRuleResponse{
+		Items: wafRules,
+		Info:  info,
+	}, nil
+}
+
+// ListAllWAFActiveRulesInput used as input for listing all WAF active rules.
+type ListAllWAFActiveRulesInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_active_rule.go [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+	FilterStatus string
+	// Limit results to active rules with the specified message.
+	FilterMessage string
+	// Limit results to active rules that represent the specified ModSecurity modsec_rule_id.
+	FilterModSedID string
+	// Include relationships. Optional, comma-separated values. Permitted values: waf_rule_revision and waf_firewall_version.
+	Include string
+}
+
+// ListAllWAFActiveRules returns the complete list of waf active rules for a given WAF id. It iterates through
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
// ListAllWAFActiveRules returns the complete list of WAF active rules for a given WAF ID. It iterates through
```

File: fastly/waf_active_rule.go [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+// all existing pages to ensure all waf active rules are returned at once.
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
// all existing pages to ensure all WAF active rules are returned at once.
```

File: fastly/waf_active_rule.go [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+		}
+		wafRules[i] = typed
+	}
+
+	return wafRules, nil
+}
+
+// BatchModificationWAFActiveRulesInput is used for active rules batch modifications.
+type BatchModificationWAFActiveRulesInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_active_rule.go [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+			Rules:            i.Rules,
+		})
+	default:
+		return nil, fmt.Errorf("operation %s not supported", i.OP)
+	}
+}
+
+// DeleteWAFActiveRulesInput used as input for removing rules from a WAF.
+type DeleteWAFActiveRulesInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_active_rule.go [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+	}
+
+	path := fmt.Sprintf("/waf/firewalls/%s/versions/%d/rules", i.WAFID, i.WAFVersionNumber)
+	_, err := c.DeleteJSONAPIBulk(path, i.Rules, nil)
+	return err
+}
+
+// DeleteAllWAFActiveRulesInput used as input for removing all rules from a WAF.
+type DeleteAllWAFActiveRulesInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 11]
```diff
@@ -224,10 +224,10 @@ func (c *Client) GetWAFVersion(i *GetWAFVersionInput) (*WAFVersion, error) {
 type UpdateWAFVersionInput struct {
 	// The Web Application Firewall's id.
 	WAFID string
-	// the Web Application Firewall's version id.
-	WAFVersionID string `jsonapi:"primary,waf_firewall"`
 	// the Web Application Firewall's version number.
-	WAFVersionNumber                 int
+	WAFVersionNumber int
+	// the Web Application Firewall's version id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's version ID.
```

File: fastly/waf_active_rule.go (hunk 4 above)
Comment (oscarDovao) [reply fastly/waf_active_rule.go]: I think we use lower case for id on similar comments

File: fastly/waf_active_rule.go (hunk 8 above)
Comment (oscarDovao) [reply fastly/waf_active_rule.go]: same as above, we use lower case for id on similar comments

File: fastly/waf_active_rule.go (hunk 9 above)
Comment (oscarDovao) [reply fastly/waf_active_rule.go]: same as above, we use lower case for id on similar comments

File: fastly/waf_active_rule.go (hunk 10 above)
Comment (oscarDovao) [reply fastly/waf_active_rule.go]: same as above, we use lower case for id on similar comments

File: fastly/waf_version.go (hunk 11 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: I've capitalised the comment, but I'd like to "id" lowercase for consistency

File: fastly/waf_active_rule.go (hunk 3 above)
Comment (oscarDovao) [reply fastly/waf_active_rule.go]: imports organised

**Comment by phamann** [comment]
@oscarDovao Whilst I agree that `id` is lowercase in other places in this code base, I'd prefer to favour the spirit of finding the code base in a better position than you found it and adhere to the community practice of [`initialisms`](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

**Comment by oscarDovao** [comment]
@phamann apologies, I thought I saw review comments arguing the use of id vs ID, but this is not the case. I'll change to ID. Thanks

File: fastly/waf_active_rule.go (hunk 4 above)
Comment (oscarDovao) [reply fastly/waf_active_rule.go]: changed on another commit

File: fastly/waf_version.go (hunk 11 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: changed on another commit

**Comment by oscarDovao** [comment]
@phamann I think I addressed all your comments. Please have a look

File: fastly/waf_active_rule.go [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+
+// BatchModificationWAFActiveRulesInput is used for active rules batch modifications.
+type BatchModificationWAFActiveRulesInput struct {
+	// The Web Application Firewall's ID.
+	WAFID string
+	// The Web Application Firewall's version number.
+	WAFVersionNumber int
+	// The list of WAF active rules (ModSecID, Status and Revision are required for upsert, ModSecID is required for delete).
+	Rules []*WAFActiveRule
+	// the batch operation to be performed (allowed operations are upsert and delete).
```
Comment (phamann) [inline fastly/waf_active_rule.go]:
```suggestion
	// The batch operation to be performed (allowed operations are upsert and delete).
```

**Pull request merged**